src/oidc_child/oidc_child.c

```diff
@@ -0,0 +1,571 @@
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <unistd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <popt.h>
+
+#include "oidc_child/oidc_child_util.h"
+
+#include "util/util.h"
+#include "util/atomic_io.h"
+
+#define IN_BUF_SIZE 4096
+static errno_t read_device_code_from_stdin(struct devicecode_ctx *dc_ctx)
+{
+    uint8_t buf[IN_BUF_SIZE];
+    ssize_t len;
+    errno_t ret;
+    char *str;
+
+    errno = 0;
+    len = sss_atomic_read_s(STDIN_FILENO, buf, IN_BUF_SIZE);
+    if (len == -1) {
+        ret = errno;
+        ret = (ret == 0) ? EINVAL: ret;
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "read failed [%d][%s].\n", ret, strerror(ret));
+        return ret;
+    }
+
+    if (len == 0 || *buf == '\0') {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing device code\n");
+        return EINVAL;
+    }
+
+    str = talloc_strndup(dc_ctx, (char *) buf, len);
+    sss_erase_mem_securely(buf, IN_BUF_SIZE);
+    if (str == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "talloc_strndup failed.\n");
+        return ENOMEM;
+    }
+    talloc_set_destructor((void *) str, sss_erase_talloc_mem_securely);
+
+    if (strlen(str) != len) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "Input contains additional data, "
+              "only JSON encoded device code expected.\n");
+        talloc_free(str);
+        return EINVAL;
+    }
+
+    clean_http_data(dc_ctx);
+    dc_ctx->http_data = str;
+
+    DEBUG(SSSDBG_TRACE_ALL, "JSON device code: [%s].\n", dc_ctx->http_data);
+
+    return EOK;
+}
+
+static errno_t set_endpoints(struct devicecode_ctx *dc_ctx,
+                             const char *device_auth_endpoint,
+                             const char *token_endpoint,
+                             const char *userinfo_endpoint,
+                             const char *jwks_uri,
+                             const char *scope)
+{
+    int ret;
+
+    dc_ctx->device_authorization_endpoint = talloc_strdup(dc_ctx,
+                                                          device_auth_endpoint);
+    if (dc_ctx->device_authorization_endpoint == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing device_authorization_endpoint.\n");
+        ret = EINVAL;
+        goto done;
+    }
+    dc_ctx->token_endpoint = talloc_strdup(dc_ctx, token_endpoint);
+    if (dc_ctx->token_endpoint == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing token_endpoint.\n");
+        ret = EINVAL;
+        goto done;
+    }
+    dc_ctx->userinfo_endpoint = talloc_strdup(dc_ctx,userinfo_endpoint);
+    if (dc_ctx->userinfo_endpoint == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing userinfo_endpoint.\n");
+        ret = EINVAL;
+        goto done;
+    }
+
+    if (jwks_uri != NULL && *jwks_uri != '\0') {
```

Comment:
Why are `jwks_uri` and `scope` handled differently than 'endpoints'?

Reply:
Those two are optional while the other are required.